src/ir/ExpressionAnalyzer.cpp

```diff
@@ -443,34 +152,125 @@ bool ExpressionAnalyzer::flexibleEqual(Expression* left,
+#define DELEGATE_FIELD_NAME(id, name) COMPARE_FIELD(name)
+#define DELEGATE_FIELD_SIGNATURE(id, name) COMPARE_FIELD(name)
+#define DELEGATE_FIELD_TYPE(id, name) COMPARE_FIELD(name)
+#define DELEGATE_FIELD_ADDRESS(id, name) COMPARE_FIELD(name)
+
+#define COMPARE_LIST(name)                                                     \
+  if (castLeft->name.size() != castRight->name.size()) {                       \
+    return false;                                                              \
+  }                                                                            \
+  for (Index i = 0; i < castLeft->name.size(); i++) {                          \
+    if (castLeft->name[i] != castRight->name[i]) {                             \
+      return false;                                                            \
+    }                                                                          \
+  }
+
+#define DELEGATE_FIELD_INT_ARRAY(id, name) COMPARE_LIST(name)
+
+#define DELEGATE_FIELD_SCOPE_NAME_DEF(id, name)                                \
+  if (castLeft->name.is() != castRight->name.is()) {                           \
+    return false;                                                              \
+  }                                                                            \
+  rightNames[castLeft->name] = castRight->name;
+
+#define DELEGATE_FIELD_SCOPE_NAME_USE(id, name)                                \
+  if (!compareNames(castLeft->name, castRight->name)) {                        \
+    return false;                                                              \
+  }
+
+#define DELEGATE_FIELD_SCOPE_NAME_USE_VECTOR(id, name)                         \
+  if (castLeft->name.size() != castRight->name.size()) {                       \
+    return false;                                                              \
+  }                                                                            \
+  for (Index i = 0; i < castLeft->name.size(); i++) {                          \
+    if (!compareNames(castLeft->name[i], castRight->name[i])) {                \
+      return false;                                                            \
+    }                                                                          \
+  }
+
+#include "wasm-delegations-fields.h"
+
+      return true;
+    }
+
+    bool compareNames(Name left, Name right) {
+      auto iter = rightNames.find(left);
+      // If it's not found, that means it was defined out of the expression
+      // being compared, in which case we can just treat it literally - it
+      // must be exactly identical.
+      if (iter != rightNames.end()) {
+        left = iter->second;
+      }
+      return left == right;
+    }
   };
 
   return Comparer().compare(left, right, comparer);
 }
 
 // hash an expression, ignoring superficial details like specific internal names
 size_t ExpressionAnalyzer::hash(Expression* curr) {
-  struct Hasher {
+  struct Hasher : Visitor<Hasher> {
```

Comment:
Where is the `Visitor` functionality used?

Reply:
Good catch, it's not... it's completely stack based + the delegation switch, so it doesn't need that. Removed.